### ISSUE TITLE
Implement better associative arrays

### ISSUE THREAD

**Issue by abensonca** [galacticusorg/galacticus#542]
The associative array ("hash") functionality in Galacticus is extremely simplistic - it is not optimized in any way. Given that we don't currently use it for any case where the number of entries becomes large this may not be a huge problem. But, better options do exist, for example [`libcdict`](https://gitlab.com/rob.izzard/libcdict) which might be easy to use (with some suitable wrappers).

**Comment by abensonca** [comment]
# Proposal: Better Associative Arrays (issue #542)

**Status:** Proposal only — no code changed, nothing committed.
**Scope guard:** This document touches the *data structure*, never the *values* it stores. See **§4 Risks** for the one place where behavior visible in output files (attribute *ordering*) could change; the recommended plan in §2/§3 explicitly avoids that.

---

## 1. Current state

### 1.1 Where the code lives

All associative-array ("hash") functionality is generated from a single
templated source file:

- `source/utility.hashes.F90` — the implementation, expanded by
  `scripts/build/preprocess.py` over a `{Type¦label}` generic block into ~15
  concrete types: `integerHash`, `integerSizeTHash`, `doubleHash`,
  `varyingStringHash`, `genericHash`, and the `rank1/rank2` `Double`/`IntegerSizeT`
  value/pointer variants.
- `source/utility.hashes.perfect.F90`, `source/utility.hashes.cryptographic.F90`
  — unrelated to issue #542 (perfect-hash code generation and crypto digests);
  out of scope here.
- Tests: `source/tests.hashes.F90`.

**Because everything is generated from one template, an algorithmic fix applies
to every concrete hash type at once.** This is the single most important fact for
sizing the work: the change surface is *one file*.

### 1.2 How it actually works

The hash is **not** a hash table. It is a pair of parallel arrays kept **sorted
by key**:

- `hashKeys(:)` — `varying_string`, maintained in sorted order.
- `hashValues(:)` — values, kept in lock-step with the keys.
- A one-entry memo (`keyPrevious`, `indexPrevious`) short-circuits repeated
  access to the same key.

Complexity of each operation today:

| Method | Mechanism | Complexity | Note |
|---|---|---|---|
| `set` (new key) | binary search for slot, then **shift array** | O(n) | shift dominates |
| `set` (existing) | binary search / memo | O(log n) | fine |
| `value(key)` | `exists()` **then** `searchArray()` | **O(n)** | does two searches; the first is linear |
| `exists(key)` | `any(hashKeys == key)` | **O(n)** | linear scan despite sorted data |
| `delete(key)` | `exists()` then shift | O(n) | |
| `value(i)`, `key(i)`, `size()` | direct index | O(1) | |
| growth | `allocatedSize += 128` (linear) | O(n²) total to build | not geometric |

Usage frequency across `source/` (method-call counts): `value` ≈ 608, `set` ≈ 486,
`exists` ≈ 188, `size` ≈ 147, `key` ≈ 99, `values` ≈ 22, `delete` ≈ 5. So the two
hottest operations — `value` and `exists` — are exactly the two that are O(n)
today.

### 1.3 Pain points (concrete, not theoretical)

1. **`value()` does redundant work.** `{Type}ValueVarStr` calls
   `{Type}ExistsVarStr` (a linear `any()` scan) and *then*, if present,
   `searchArray` (a binary search) — two passes, the first linear, to retrieve
   one element (`source/utility.hashes.F90:359-369`). Since the keys are already
   sorted, this should be a *single* O(log n) binary search.
2. **`exists()` is linear** (`any(self%hashKeys(...) == key)`,
   `source/utility.hashes.F90:211`) even though the keys are sorted and a binary
   search is right there.
3. **Linear growth increment** of `hashSizeIncrement=128`
   (`source/utility.hashes.F90:113`, `:445`). Building an N-entry hash triggers
   ~N/128 reallocations, each copying O(N) — O(N²) build cost for large N.
4. **Insert/delete shift the whole tail** of the array (O(n)). Unavoidable for a
   contiguous sorted array; only matters if a hash is built incrementally and
   large.
5. **Every comparison is a `varying_string` comparison.** No cheap integer
   pre-check (e.g. a stored key-hash) to reject mismatches before the string
   compare.

The issue itself notes hashes are "extremely simplistic… not optimized in any
way" and that we "don't currently use it for any case where the number of
entries becomes large." That last clause matters: **today this is a latency
papercut, not a scaling wall.** It justifies an *incremental* fix, not a rewrite.

---

## 2. Proposed approach

**Recommendation: an in-place algorithmic improvement to `utility.hashes.F90`,
keeping the sorted-array storage and the exact external interface.** Defer (do
not adopt now) the external-library route from the issue.

### 2.1 Why not `libcdict` (the issue's suggestion) — at least not first

`libcdict` is a real hash-table library and would give O(1) average ops. But for
this codebase, as a *first* step, it is the wrong trade:

- **It changes iteration order.** A hash table iterates in bucket order; our
  sorted arrays iterate in sorted-key order. The outputter writes metadata as
  `do i=1,size; writeAttribute(value(i), key(i))`
  (`source/merger_trees.outputter.standard.F90:758-828`) and nBody operators do
  similar. Changing iteration order changes the *order attributes are written to
  HDF5* — a file-level diff even though every value is identical. See §4.
- **New external/C dependency** (build, packaging, the "with some suitable
  wrappers" caveat) and ISO_C_Binding marshaling of `varying_string` keys and of
  our `class(*)` / rank-1/rank-2 value variants — non-trivial for the generic
  and array-valued instances.
- **No demonstrated need for O(1) yet** (issue: entries never become large).

The library is a reasonable *future* option once (a) a real large-N use case
appears and (b) ordering is decoupled from storage (§3, optional last step). The
incremental fix below captures the bulk of the realistic win at a fraction of the
risk and keeps that door open.

### 2.2 The minimal change

Three internal edits to `source/utility.hashes.F90`, all behavior-preserving at
the interface:

**(A) Single-pass `value()`** — replace exists-then-search with one binary
search:

```fortran
! {Type}ValueVarStr  — sketch (replaces lines ~359-369)
if (key == self%keyPrevious) then
   {Type}ValueVarStr {Type¦assignment} self%hashValues(self%indexPrevious)%object
else
   iKey = self%lookup(key)            ! single binary search, 0 if absent
   if (iKey > 0) then
      {Type}ValueVarStr {Type¦assignment} self%hashValues(iKey)%object
      self%keyPrevious = key ; self%indexPrevious = iKey   ! refresh memo
   else
      {Type}ValueVarStr {Type¦assignment} {Type¦null}
      call Error_Report('key '''//char(key)//''' does not exist in hash'//...)
   end if
end if
```

**(B) Binary-search `exists()`** — share the same private `lookup` helper:

```fortran
logical function {Type}ExistsVarStr(self,key)
   {Type}ExistsVarStr = self%lookup(key) > 0
end function
```

where `lookup` wraps the existing `searchArray` plus an exact-match check:

```fortran
! private; returns matching index or 0 (absent). O(log n).
integer(c_size_t) function {Type}Lookup(self,key) result(iKey)
   if (self%elementCount == 0) then
      iKey = 0 ; return
   end if
   iKey = searchArray(self%hashKeys(1:self%elementCount), key)   ! existing routine
   if (iKey < 1 .or. iKey > self%elementCount) then
      iKey = 0
   else if (.not. (self%hashKeys(iKey) == key)) then
      iKey = 0
   end if
end function
```

`set` and `delete` already binary-search for the slot; they would route their
existence check through the same `lookup` for consistency, but their O(n) shift
is unchanged (intentionally — see §2.3).

**(C) Geometric growth** — change the grow step from additive to multiplicative
(`source/utility.hashes.F90:445`):

```fortran
self%allocatedSize = max(self%allocatedSize*2, hashSizeIncrement)   ! was + hashSizeIncrement
```

This turns O(N²) build into amortized O(N) with no interface change. (Keep
`hashSizeIncrement` as the floor for the first allocation.)

### 2.3 What is deliberately *not* changed

- **External interface**: every public method signature, `assignment(=)`, the
  constructor, the generic char/varstr overloads — unchanged.
- **Storage layout & iteration order**: still a sorted array, so `keys()`,
  `values()`, `key(i)`, `value(i)` return entries in the *same order as today*.
  This is what keeps output files byte-identical (§4).
- **O(n) insert/delete**: left as-is. Fixing it means abandoning the contiguous
  sorted array (a tree or hash table), which is precisely the larger change we
  are deferring. The geometric-growth fix (C) already removes the worst
  build-time pathology without touching layout.

### 2.4 Optional, larger follow-up (separate proposal, not now)

If a genuine large-N hot path emerges, the clean evolution is to **decouple
ordering from storage**: switch internal storage to an open hash table (or
`libcdict`) for O(1) `set`/`value`/`exists`, and make `keys()`/`values()`/`key(i)`
return entries **sorted on demand** so the external ordering contract — and thus
output files — is preserved. That isolates the only behavioral risk into a
sort-on-iteration step. Out of scope for issue #542's first pass.

---

## 3. Migration plan

The work is intentionally staged so each step is independently verifiable and
reversible. **No call sites change** in steps 1–3 (that is the point of keeping
the interface fixed).

| Step | Change | Files | Verify |
|---|---|---|---|
| 0 | Baseline | — | Build clean; run `tests.hashes` and the full regression suite; **archive reference output HDF5 files** for the diff in §4. |
| 1 | Add private `lookup` helper; route `exists()` through it | `utility.hashes.F90` | `tests.hashes` passes; add unit cases for empty hash, key below/above range, single element, exact vs. near-miss key. |
| 2 | Rewrite `value(key)` to single-pass via `lookup` (keep memo) | `utility.hashes.F90` | `tests.hashes` passes incl. the "missing key errors" path; assert error is still raised for absent key. |
| 3 | Geometric growth in `set` | `utility.hashes.F90` | `tests.hashes` passes; add a stress test inserting e.g. 10⁴ keys in adversarial (reverse-sorted) order, assert all retrievable and `size` correct. |
| 4 | Full regression run + **bit-for-bit output diff** vs. step-0 references | none | See §4 verification. |
| 5 (optional) | Micro-benchmark before/after | new throwaway driver | Confirm `value`/`exists` scale ~log n; build time ~linear. Not required to land. |

Ordering rationale: steps 1→3 are each a strict internal refactor with the test
suite as the gate, smallest first. The behavioral safety net (step 4) runs once
all three are in.

**Call sites:** none are *edited*. The ~37 files that `use Hashes` and the
~1500 method calls all keep working unchanged — that is the design goal. (For
context, heaviest users are `merger_trees.construct.read.importer.galacticus.F90`,
`utility.input_parameters.F90`, and the nBody operators; none need touching.)

---

## 4. Risks

### 4.1 Numerical / physics results — **assessed: no change, with one guard**

The hash stores key→value mappings; this refactor changes only *how a value is
located and how the backing array grows*, never the stored value. `value(key)`
returns the same bits for the same key. **No physics computation reads anything
order-dependent from a hash** in the steps-1–3 design, because iteration order is
preserved (storage stays sorted).

> **FLAG (the one place output could move):** the standard outputter and several
> nBody operators iterate a hash by index and write each entry as an HDF5
> *attribute* (`merger_trees.outputter.standard.F90:758-828`;
> `nBody.operator.add_attributes.F90`). If a future variant (the §2.4 follow-up,
> or `libcdict`) changes iteration order, the **order attributes are written
> changes**. Attribute *values* and names are identical and readers look them up
> by name, so this is **not** a numerical/physics change — but it *is* a
> byte-level change in output files that regression tests doing exact file
> comparison would catch. **The recommended steps-1–3 plan keeps storage sorted
> and therefore keeps this order unchanged.** This risk applies only if/when the
> optional §2.4 work is taken up, at which point the mitigation is the
> sort-on-iteration step described there.

**Verification for step 4:** run the standard regression models before and after,
and diff the output HDF5 (`h5diff`, and a structural compare of attribute
ordering). Expect *identical* files. If anything differs, stop — it means an
ordering assumption leaked in.

### 4.2 Correctness risks in the refactor itself

- **Off-by-one / boundary in `lookup`.** `searchArray` returns `0` below range
  and `n+1` above; the exact-match guard in `lookup` must handle both plus the
  empty hash. Covered by the step-1 unit cases.
- **Memo coherence.** `value()` must refresh `keyPrevious/indexPrevious` on a
  hit, and `set`/`delete` already invalidate it; the rewrite must not drop those
  invalidations (delete sets them to `''/-1`).
- **Pointer-valued and `generic` instances.** The `Ptr` and `class(*)` variants
  use `=>` assignment and special deallocate paths in `set`
  (`utility.hashes.F90:424-429`); the `value`/`exists` changes don't touch
  assignment, but the test matrix should include `genericHash` (already in
  `tests.hashes.F90`) and at least one `rank1DoublePtrHash`.

### 4.3 Performance risk

- **Geometric growth raises peak memory** (up to ~2× transient during reallocate,
  and up to 2× steady-state slack vs. the old +128). For the current small hashes
  this is negligible; keeping `hashSizeIncrement` as the floor avoids waste on
  tiny hashes. If any hash is known to be huge, a `trim`/`shrink_to_fit` method
  could be added later — not needed now.
- **No expected regression** on the hot paths: `value`/`exists` go from O(n) to
  O(log n); the memo fast-path is retained, so repeated same-key access (common
  in input-parameter lookups) is unchanged.

### 4.4 Scope / process risks

- Change is confined to one generated file; low blast radius. The chief way to
  *expand* scope accidentally is to start "fixing" insert/delete O(n) or adopting
  `libcdict` — both explicitly deferred here.
- Per task constraints: this is a proposal; **no refactor performed, nothing
  committed**, and the single output-affecting scenario is flagged above rather
  than enacted.

---

## 5. Recommendation in one line

Do the three internal edits (single-pass `value`, binary-search `exists`,
geometric growth) to `source/utility.hashes.F90`, gated by `tests.hashes` plus a
bit-for-bit output diff; keep the sorted-array layout so iteration order — and
every output file — is unchanged. Revisit `libcdict` only if a real large-N use
case appears, and only together with sort-on-iteration to protect output
ordering.
